From 7a56b9de3a8ca4ed4cf106ece082818ff11afe17 Mon Sep 17 00:00:00 2001
From: Jason Hung <jhung@globalscaletechnologies.com>
Date: Wed, 17 Jul 2024 17:13:12 +0800
Subject: [PATCH] board: mediatek: add Globalscale MT510SS evk board

Add Globalscale MT510SS EVK board support.
This EVK evaluation board uses MediaTek's MT8370 SoC (Genio-510).
---
 arch/arm/dts/Makefile            |   3 +-
 arch/arm/dts/gti-mt510ss-evk.dts | 365 +++++++++++++++++++++++++++++++
 2 files changed, 367 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm/dts/gti-mt510ss-evk.dts

diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index e10c5ace83..8f68af6765 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -1244,7 +1244,8 @@ dtb-$(CONFIG_ARCH_MEDIATEK) += \
 	genio-510-evk.dtb \
 	genio-700-evk.dtb \
 	genio-1200-evk.dtb \
-	genio-1200-evk-ufs.dtb
+	genio-1200-evk-ufs.dtb \
+	gti-mt510ss-evk.dtb
 
 dtb-$(CONFIG_ARCH_NPCM7xx) += nuvoton-npcm750-evb.dtb
 dtb-$(CONFIG_XEN) += xenguest-arm64.dtb
diff --git a/arch/arm/dts/gti-mt510ss-evk.dts b/arch/arm/dts/gti-mt510ss-evk.dts
new file mode 100644
index 0000000000..1ed76ca57f
--- /dev/null
+++ b/arch/arm/dts/gti-mt510ss-evk.dts
@@ -0,0 +1,365 @@
+// SPDX-License-Identifier: GPL-2.0 OR MIT
+/*
+ * Copyright (C) 2023 MediaTek Inc.
+ * Author: Macpaul Lin <macpaul.lin@mediatek.com>
+ */
+
+/dts-v1/;
+
+#include <config.h>
+#include "mt8370.dtsi"
+
+/ {
+	model = "Globalscale MT510-SS EVK";
+	compatible = "globalscale,mt510ss-evk",
+				 "mediatek,genio-510-evk",
+				 "mediatek,mt8370";
+
+	memory@40000000 {
+		/* 4GB */
+		device_type = "memory";
+		reg = <0 0x40000000 1 0x00000000>;
+	};
+
+	firmware: firmware {
+		optee {
+			compatible = "linaro,optee-tz";
+			method = "smc";
+		};
+
+		psci: psci {
+			compatible = "arm,psci-1.0";
+			method = "smc";
+		};
+	};
+
+	reserved-memory {
+		#address-cells = <2>;
+		#size-cells = <2>;
+		ranges;
+
+		/* 2 MiB reserved for ARM Trusted Firmware (BL31) */
+		bl31_secmon_reserved: secmon@54600000 {
+			no-map;
+			reg = <0 0x54600000 0x0 0x200000>;
+		};
+
+		/* 12 MiB reserved for OP-TEE (BL32)
+		 * +-----------------------+ 0x43e0_0000
+		 * |      SHMEM 2MiB       |
+		 * +-----------------------+ 0x43c0_0000
+		 * |        | TA_RAM  8MiB |
+		 * + TZDRAM +--------------+ 0x4340_0000
+		 * |        | TEE_RAM 2MiB |
+		 * +-----------------------+ 0x4320_0000
+		 */
+		optee_reserved: optee@43200000 {
+			no-map;
+			reg = <0 0x43200000 0 0x00c00000>;
+		};
+
+		dsi_reserved: dsi@60000000 {
+			reg = <0 0x60000000 0 0x02000000>;
+			no-map;
+		};
+	};
+
+	chosen {
+		stdout-path = &uart0;
+	};
+
+	/* dummy eMMC 1.8v power supply */
+	reg_1p8v: regulator-1p8v {
+		compatible = "regulator-fixed";
+		regulator-name = "fixed-1.8V";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+	/* dummy eMMC 3.3v power supply */
+	reg_3p3v: regulator-3p3v {
+		compatible = "regulator-fixed";
+		regulator-name = "fixed-3.3V";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+    /* 3.3v power for wifi module */
+	wcn_3v3_en: wcn-3v3-en {
+		compatible = "regulator-fixed";
+		regulator-name = "wcn-3v3-en";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&gpio 74 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+	usbp0_vbus3v3: usbp0-vbus3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "usbp0-vbus3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&gpio 84 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+	usbhub_3v3: usbhub-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "usbhub-3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&gpio 112 GPIO_ACTIVE_HIGH>;
+		startup-delay-us = <10000>;
+		enable-active-high;
+	};
+
+	usbhub_reset: usbhub-reset {
+		compatible = "regulator-fixed";
+		regulator-name = "usbhub-reset";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		gpio = <&gpio 7 GPIO_ACTIVE_HIGH>;
+		enable-active-low;
+	};
+
+	dsi0: dsi0{
+		compatible = "mediatek,mt8188-dsi";
+		reg = <0 0x60000000 0 0x02000000>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&panel_pins_default>;
+		reset-gpios = <&gpio 25 GPIO_ACTIVE_HIGH>;
+		dcdc-gpios = <&gpio 45 GPIO_ACTIVE_HIGH>;
+		enable-gpios = <&pca9554 7 GPIO_ACTIVE_HIGH>;
+		status = "okay";
+	};
+};
+
+&watchdog {
+	status = "okay";
+};
+
+&uart0 {
+	status = "okay";
+};
+
+&mmc0 {
+	bus-width = <8>;
+	max-frequency = <200000000>;
+	cap-mmc-highspeed;
+	mmc-hs200-1_8v;
+	mmc-hs400-1_8v;
+	hs400-ds-delay = <0x1481b>;
+	cap-mmc-hw-reset;
+	vmmc-supply = <&reg_3p3v>;
+	vqmmc-supply = <&reg_1p8v>;
+	non-removable;
+	status = "okay";
+};
+
+&pio {
+	panel_pins_default: panel-pins-default {
+		panel-dcdc {
+			pinmux = <PINMUX_GPIO45__FUNC_B_GPIO45>;
+			output-low;
+		};
+
+		panel-rst {
+			pinmux = <PINMUX_GPIO25__FUNC_B_GPIO25>;
+			output-high;
+		};
+
+		pins-bl {
+			pinmux = <PINMUX_GPIO30__FUNC_O_DISP_PWM1>;
+		};
+	};
+
+	dp_pin_default: dp-pin-default {
+		dp-panel-vcc {
+			pinmux = <PINMUX_GPIO14__FUNC_B_GPIO14>;
+			output-low;
+		};
+
+		dp-panel-hpd {
+			pinmux = <PINMUX_GPIO46__FUNC_I0_DP_TX_HPD>;
+			bias-pull-up;
+		};
+	};
+
+	i2c0_pin: i2c0-pin {
+		pins-bus {
+			pinmux = <PINMUX_GPIO56__FUNC_B1_SDA0>,
+				 <PINMUX_GPIO55__FUNC_B1_SCL0>;
+			input-enable;
+			bias-pull-up;
+		};
+	};
+
+	i2c1_pin: i2c1-pin {
+		pins-bus {
+			pinmux = <PINMUX_GPIO58__FUNC_B1_SDA1>,
+				 <PINMUX_GPIO57__FUNC_B1_SCL1>;
+			input-enable;
+			bias-pull-up;
+		};
+	};
+
+	i2c2_pin: i2c2-pin {
+		pins-bus {
+			pinmux = <PINMUX_GPIO60__FUNC_B1_SDA2>,
+				 <PINMUX_GPIO59__FUNC_B1_SCL2>;
+			input-enable;
+			bias-pull-up;
+		};
+	};
+
+	i2c3_pin: i2c3-pin {
+		pins-bus {
+			pinmux = <PINMUX_GPIO62__FUNC_B1_SDA3>,
+				 <PINMUX_GPIO61__FUNC_B1_SCL3>;
+			input-enable;
+			bias-pull-up;
+		};
+	};
+
+	i2c4_pin: i2c4-pin {
+		pins-bus {
+			pinmux = <PINMUX_GPIO64__FUNC_B1_SDA4>,
+				 <PINMUX_GPIO63__FUNC_B1_SCL4>;
+			input-enable;
+			bias-pull-up;
+		};
+	};
+
+	i2c5_pin: i2c5-pin {
+		pins-bus {
+			pinmux = <PINMUX_GPIO66__FUNC_B1_SDA5>,
+				 <PINMUX_GPIO65__FUNC_B1_SCL5>;
+			input-enable;
+			bias-pull-up;
+		};
+	};
+
+	i2c6_pin: i2c6-pin {
+		pins-bus {
+			pinmux = <PINMUX_GPIO68__FUNC_B1_SDA6>,
+				 <PINMUX_GPIO67__FUNC_B1_SCL6>;
+			input-enable;
+			bias-pull-up;
+		};
+	};
+};
+
+&i2c0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c0_pin>;
+	clock-frequency = <400000>;
+	status = "okay";
+};
+
+&i2c1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c1_pin>;
+	clock-frequency = <400000>;
+	status = "okay";
+};
+
+&i2c2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c2_pin>;
+	clock-frequency = <400000>;
+	status = "okay";
+
+	pca9554: gpio@3b {
+		compatible = "nxp,pca9554";
+		reg = <0x3b>;
+		status = "okay";
+	};
+};
+
+&i2c3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c3_pin>;
+	clock-frequency = <400000>;
+	status = "okay";
+};
+
+&i2c4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c4_pin>;
+	clock-frequency = <400000>;
+	status = "okay";
+};
+
+&i2c5 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c5_pin>;
+	clock-frequency = <400000>;
+	status = "okay";
+};
+
+&i2c6 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c6_pin>;
+	clock-frequency = <400000>;
+	status = "okay";
+};
+
+&u3phy0 {
+	status = "okay";
+};
+
+&u3phy1 {
+	status = "okay";
+};
+
+&u3phy2 {
+	status = "okay";
+};
+
+&u2port0 {
+	status = "okay";
+};
+
+&u2port1 {
+	status = "okay";
+};
+
+&u3port1 {
+	status = "okay";
+};
+
+&u2port2 {
+	status = "okay";
+};
+
+/* usb2.0 p0 */
+&usb {
+	vbus-supply = <&usbp0_vbus3v3>;
+	status = "okay";
+};
+
+/* usb2 p0 */
+&ssusb {
+	mediatek,force-vbus;
+	maximum-speed = "high-speed";
+	dr_mode = "peripheral";
+	status = "okay";
+};
+
+/* usb3.0 p1 to usb3 hub */
+&xhci1 {
+	vusb33-supply = <&usbhub_3v3>;
+	vbus-supply = <&usbhub_reset>;
+	status = "okay";
+};
+
+/* usb2 p2 to usb2 hub */
+&xhci2 {
+	status = "okay";
+};
